### PR TITLE
Migrate from TSLint to ESLint

### LINKED ISSUE #115: Migrate from tslint to eslint with typescript-eslint
ESLint is getting full typescript support with [`typescript-eslint`](https://github.com/typescript-eslint/typescript-eslint), replacing TSLint.
As you can see in their readme the project is backed by both the TSLint developer and the TypeScript Team, so this is going to be the official linting solution.

### PR DESCRIPTION
**Pull request by tcdejong**
Closes #115. Used the [tslint-to-eslint-config](https://github.com/typescript-eslint/tslint-to-eslint-config) tool and slightly modified the resulting configuration to reintroduce missing rules. 

Minor change to `main.ts` so that eslint could actually find the `ErrorWrapper`. 

This is a big change so I'd argue it requires thorough testing and bumping up the version number. 
It's also possible I may have missed certain rules. I tested it with a few of my old modules and I think we've got 90% of it. 

I tested the repository by downloading the source as usual, running `npm install` and `npm run-script build`. Everything seems to be working.

### CODE REVIEW

**Comment by tcdejong** [comment]
After messing around more I think there are a bunch of redundant rules from the config tool, and some still being missing. When I have time I might mess around with the rules more. 

We could also create a feature branch on the main repo and work on it there before merging the transition.

**Comment by resir014** [comment]
> After messing around more I think there are a bunch of redundant rules from the config tool, and some still being missing. When I have time I might mess around with the rules more.
> 
> We could also create a feature branch on the main repo and work on it there before merging the transition.

Sure, I can do that when I have the time. @tcdejong

**Comment by djD-REK** [comment]
Any update or can I help?

**Comment by tcdejong** [comment]
The easiest way is if you make a fork and see if everything is working, and check if the ESLint rules are complete and without redundancy or unnecessary rules. 

Personally I won't have time to work on this anymore, or anything screeps related for that matter.

**Comment by djD-REK** [comment]
@tcdejong  Thanks for the comment, I'll work on it as I'm getting familiarized with the package.  Thanks for your work on this! 🙇

**Comment by lf-** [comment]
`@typescript-eslint/no-use-before-define` should possibly be `['error', {functions: false}]` as it may genuinely catch some bugs if enabled, but the default config is broken (seriously, strictly enforcing function ordering when it doesn't matter since they're hoisted?).

**Comment by resir014** [comment]
@lf- Just got around to this. I've added the rules as you suggested, alongside some other rules.

I've also upgraded some dependencies + resolved conflicts with master. This is good to go, for the redundant rules we can remove as we go along, or periodically move them over to more standardised rulesets like Airbnb, etc.

**Review by resir014 — APPROVED** [review]

File: .eslintrc.js [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+      {
+        hoist: "all"
+      }
+    ],
+    "no-throw-literal": "error",
+    "no-trailing-spaces": "off",
+    "no-undef-init": "error",
+    "no-underscore-dangle": "warn",
+    "no-unsafe-finally": "error",
+    "no-unused-expressions": "error",
```
Comment (lf-) [inline .eslintrc.js]: `no-unused-expressions` is broken with TypeScript syntax and you should use the `@typescript-eslint/no-unused-expressions` equivalent.

**Review by lf- — CHANGES_REQUESTED** [review]

File: .eslintrc.js (hunk 1 above)
Comment (resir014) [reply .eslintrc.js]: @lf- done.

**Comment by BlueskyFR** [comment]
Why aren't the recommended typescript rules used by default? Why so much cutomization in .eslintrc.js?

**Comment by BlueskyFR** [comment]
Also, a new version of these plugin was released yesterday, we should consider upgrading:
```json
"@typescript-eslint/eslint-plugin": "^3.7.0",
"@typescript-eslint/parser": "^3.7.0",
"@typescript-eslint/typescript-estree": "^3.7.0",
```

**Comment by resir014** [comment]
> Why aren't the recommended typescript rules used by default? Why so much cutomization in .eslintrc.js?

@BlueskyFR As per OP + as mentioned earlier, this was migrated from the old TSLint rules to avoid too much breakage on the users' side should they upgrade. There are still a bunch of redundant rules that needs removing, but we'll adjust as we go.

**Comment by BlueskyFR** [comment]
> > Why aren't the recommended typescript rules used by default? Why so much cutomization in .eslintrc.js?
> 
> @BlueskyFR As per OP + as mentioned earlier, this was migrated from the old TSLint rules to avoid too much breakage on the users' side should they upgrade. There are still a bunch of redundant rules that needs removing, but we'll adjust as we go.

Yes sure but I mean, why was there initially (before the eslint migration) so much customization in the tslint rules?
Isn't it better to use the recommended default set without too much customization to let everyone decide what they prefer instead of deciding for them?

**Comment by resir014** [comment]
> Yes sure but I mean, why was there initially (before the eslint migration) so much customization in the tslint rules?

@BlueskyFR There weren't even that many rules before, and we did use the recommended TSLint settings: https://github.com/screepers/screeps-typescript-starter/blob/43e4f741f8fc027eaae70b2ca62bfb2c96b7270b/tslint.json

**Comment by resir014** [comment]
We'll merge this for now to let ourselves adjust ESLint rule changes as we go. These things will be done on a separate PR:

- Upgrading TS parsers
- Removing redundant/migrated rules
- Extending other recommended rulesets (possibly Airbnb) without breaking important rulesets

**Pull request merged**